Salesforce 10 User Guide (#576)
mulesoft/docs-connectors: modules/ROOT/pages/salesforce/salesforce-connector-100-config-topics.adoc

= Salesforce Additional Configuration - Mule 4
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> :imagesdir: ../../assets/images/

Reviewer: don't need Line 5 per Fer

Configuration and usage information.

[[customevents]]
== Custom Event Notifications

> The Salesforce connector provides two operations that are useful for getting custom event notifications. These notifications pertain to general events that are not tied to Salesforce data changes.

Reviewer: First use in topic Anypoint Connector for Salesforce (Salesforce Connector)

Reviewer: What are the two operations? Next sentence mentions notifications. Are those the same as the aforementioned operations?

. Create a streaming channel with the Publish Streaming Channel operation.
+

> A StreamingChannel is a special Salesforce object that represents a channel used for notifying listeners of generic Streaming API events.

Reviewer: ```suggestion
A `StreamingChannel` is a special Salesforce object that represents a channel used to notify listeners of generic streaming API events.
```

+

> Note that you can also create a streaming channel through the Salesforce or through Workbench.

Reviewer: You can create a streaming channel through "the Salesforce"? Through Salesforce Connector? And what is Workbench?

+
. Subscribe to the channel through the `Subscribe Channel Listener` operation.
+

> The Salesforce connector converts the custom events in your streaming channel to Mule events and dispatches them to your flows.

Reviewer: Salesforce Connector

[[eventsandtopics]]
== Events and Topics

Your application can receive events by subscribing to a Salesforce topic.

> Each event that travels through your flows contains information about the Salesforce data that has changed, how it changes, and when. The connector parses this information and sends you information that a flow can work with.

Reviewer: ```suggestion
Each event that travels through your flows contains information about the Salesforce data that has changed, how it changed, and when it changed. The connector parses this information and sends you information that a flow can work with.
```

Inbound properties of events:

* `payload`
* `createdDate`
* `replayId`

Salesforce stores events for 24 hours, so you can retrieve stored events during that retention window. A subscriber (to a topic or channel) can retrieve events at any time and is not restricted to listening to events at the time they are sent.

Each broadcast event is assigned a numeric ID. IDs are incremented and not guaranteed to be contiguous for consecutive events. Each ID is guaranteed to be higher than the ID of the previous event. For example, the event following the event with ID 999 can have an ID of 1025. The ID is unique for the organization and the channel. The IDs of deleted events are not reused.

See also <<Receive Data From Salesforce>> for event processing
when streaming data to an application from Salesforce.

Sections:

* <<Receive Events for a Topic>>
* <<Subscribe to a Topic>>
* <<Replay a Topic>>

[[eventsreceive]]
== Receive Events for a Topic

> Before you can receive events for Salesforce changes that are associated with a topic, you must first create a topic (a PushTopic). A PushTopic is a special object in Salesforce that binds a name (the topic's name) and Salesforce Object Query Language (SOQL) query together. Once a PushTopic is created, you can subscribe to it by using its name.

Reviewer: ```suggestion
Before you can receive events for Salesforce changes that are associated with a topic, you must first create a PushTopic. A PushTopic is a special object in Salesforce that binds the topic's name and a Salesforce Object Query Language (SOQL) query together. After a PushTopic is created, you can subscribe to it by using its name.
```

Author: Changed since your edit,

pushTopic.ApiVersion = 23.0;
pushTopic.Name = 'AllAccounts';
pushTopic.Description = 'All records for the Account object';
pushTopic.Query = 'SELECT Id, Name FROM Account';
insert pushTopic;
System.debug('Created new PushTopic: '+ pushTopic.Id);
----

[[topicsub]]
== Subscribe to a Topic

After you create a Salesforce topic, you can start receiving events by subscribing to the topic. To do so, add the Subscribe Topic Listener (`subscribe-topic-listener`) or Replay Topic Listener (`replay-topic-listener`) as an input source trigger to your flow. The trigger acts as an inbound endpoint. Every time the subscription receives an event, the trigger executes the rest of the flow in your Mule app. In the case of the XML example below, it prints a message to the log at INFO level.

In XML, you use `subscribe-topic-listener` or `replay-topic-listener` as the trigger:

[source,xml,linenums]
----
<flow name="accountUpdatesSubscription">
    <!-- INBOUND ENDPOINT -->
    <sfdc:subscribe-topic-listener topic="AccountUpdates"/>
    <!-- REST OF YOUR FLOW -->
    <logger level="INFO" message="Received an event for Salesforce Object ID #[map-payload:Id]"/>
</flow>
----

> NOTE: When subscribing to a topic that was not previously published in Salesforce, the subscription is successful. When the topic is later published, the user who is already subscribed to it does not receive notifications regarding that topic. The user has to resubscribe after the topic creates.

Reviewer: ```suggestion
NOTE: When to a topic that was not previously published in Salesforce is subscribed to, the subscription is successful. When the topic is later published, you must resubscribe to the topic to receive notifications for the topic. 
```

Author: Wording already changed.

image::salesforce/salesforce-studio-subscribe-streaming-channel.png[subscribe streaming channel]

Salesforce connector: Listens to notifications on a topic and feeds the data into the flow.

See also: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/intro_stream.htm[Salesforce Streaming API]

Streaming channels provide notifications to subscribers that are not limited to record-based events. You can use the Salesforce connector to work with Salesforce streaming channels.

[[streamingchannel]]
== Create a Streaming Channel to Receive Data From Salesforce

You must have the proper Salesforce Streaming API permissions enabled in your organization.

. Log into your Salesforce Developer Edition organization.
. Under *All Tabs (+)*, select *Streaming Channels*.
. On the *Streaming Channels* tab, select *New* to create a new streaming channel.
. Enter `/u/notifications/ExampleUserChannel` in the *Streaming Channel Name* field, and an optional description.

> You can either use the *Create* operation or the exclusive `publish-streaming-channel` operation as follows:

Reviewer: Two different tagging styles on the operations. I think they should both be back ticks

Every time a subscription to `/u/TestStreaming` receives an event, it executes the rest of the flow. In the case of this example, it logs a message at INFO level.

NOTE: The *Streaming Channel* field of the Subscribe Channel Listener operation does not display change events that are available in the Salesforce environment because this information is not provided in the Salesforce API. However, your connector can subscribe to a streaming channel. For example, to subscribe to the `All Change Events` channel, use `/data/ChangeEvents` as the channel name to which you subscribe. See https://developer.salesforce.com/docs/atlas.en-us.change_data_capture.meta/change_data_capture/cdc_subscribe_channels.htm[Subscription Channels] in the Salesforce Change Data Capture Developer Guide.

=== Streaming Channel Inbound Properties

This information gets passed along as inbound properties:

> * `channel` - Maps to the Channel JSON property.

Reviewer: No periods on these list items

Author: They are sentences now
